### ISSUE TITLE
How do I find shortest path via Java API and can I limit the search range?

### ISSUE THREAD

**Issue by acmeguy** [orientechnologies/orientdb#5771]

**Comment by acmeguy** [comment]
Currently I'm using this approach:

"select shortestPath("+ sourceId + ", "+ destinationId +")"

But I'm only interested in certain depth and this seems to exhaust every possibility (which is a overkill for me)

**Comment by lvca** [comment]
For search range what do you mean? Can you do a concrete example of what you'd like to get?

**Comment by acmeguy** [comment]
I would like to limit the number of hops/degrees the shortest path can go (I'm only interested to know if a relationship is within X degrees)

**Comment by acmeguy** [comment]
Meaning: I need to know if the two entities are linked over, for example, 4 degrees or less. I do not need a complete/exhaustive search to happen that returns a path of 20 hops/degrees and I'm trying to avoid the time that it takes. We have ~40 million entities in our db and density is an issue so I would also like to limit the Edge types that can be traversed in search of the shortest path.

**Comment by lvca** [comment]
@luigidellaquila WDYT about supporting a limit on this function? Or is there any other way, for example with the MATCH to have the same goal?

**Comment by luigidellaquila** [comment]
I think shortestPath() function can be enhanced with many parameters, one of them is for sure the max depth.

MATCH statement is not optimized (yet) to calculate shortest paths, you can write a query to get the same result but the elaboration time will be much higher than shortest path

Thanks

Luigi

**Comment by acmeguy** [comment]
Ok, so this is not an option.  The time spent will be determined by the density of the graph so I would really appreciate a query that would achieve this. The only other option is for us to create our own traversing using the Java API and I would like to avoid that at all cost.

**Comment by luigidellaquila** [comment]
Ok, let me see what I can do on shortestPath(), maybe I can quickly enhance it

Thanks

Luigi

**Comment by luigidellaquila** [comment]
Hi @acmeguy 

I think I have a solution for this, I'm running all the tests and then I'll push it to 2.1.x branch

Thanks

Luigi

**Comment by acmeguy** [comment]
Great stuff, thank you :)

**Comment by luigidellaquila** [comment]
Ok, pushed on both 2.1.x and develop branches. 

Now `shortestPath()` accepts a new (fifth) parameter that is a map and can contain multiple options.

The only supported option is "maxDepth" right now:
```
SELECT shortestPath(#8:32, #8:10, null, null, {"maxDepth": 5})
```
Here are full docs

https://github.com/orientechnologies/orientdb-docs/blob/master/SQL-Functions.md#shortestpath

The enhancement will be released with 2.1.12

Thanks

Luigi

**Comment by acmeguy** [comment]
thank you!
